java/ql/lib/semmle/code/java/security/InsufficientKeySize.qll

```diff
@@ -3,32 +3,76 @@
 private import semmle.code.java.security.Encryption
 private import semmle.code.java.dataflow.DataFlow
 private import semmle.code.java.security.internal.EncryptionKeySizes
+import codeql.util.Either
+
+/** A minimum recommended key size for some algorithm. */
+abstract class MinimumKeySize extends int {
+  bindingset[this]
+  MinimumKeySize() { any() }
+
+  /** Gets a textual representation of this element. */
+  string toString() { result = super.toString() }
+}
+
+/**
+ * A class of algorithms for which a key size smaller than the recommended key
+ * size might be embedded in the algorithm name.
+ */
+abstract class AlgorithmKind extends string {
+  bindingset[this]
+  AlgorithmKind() { any() }
+
+  /** Gets a textual representation of this element. */
+  string toString() { result = super.toString() }
+}
+
+/**
+ * A key size that is greater than the tracked value and equal to the minimum
+ * recommended key size for some algorithm, or a kind of algorithm for which the
+ * tracked string indicates a too small key size.
+ */
+final class KeySizeState = Either<MinimumKeySize, AlgorithmKind>::Either;
 
 /** A source for an insufficient key size. */
 abstract class InsufficientKeySizeSource extends DataFlow::Node {
   /** Holds if this source has the specified `state`. */
-  predicate hasState(DataFlow::FlowState state) { state instanceof DataFlow::FlowStateEmpty }
+  abstract predicate hasState(KeySizeState state);
```

Comment:
Yes, but such a hypothetical source wouldn't make much sense. Its default state would not match any sink, so it wouldn't work. Hence, I find it unlikely that this extension point has been used in the wild. Really, as the query is written, we _need_ a key size state to be specified for the sources and the sinks, and with the way that I've reimplemented the logic I found it hard to provide a straight-forward deprecation path from the old interface.

I guess we could put in some work provide a better deprecation path by renaming the `hasState` predicate and providing a translation from the old to the new, but it's not going to be neither pretty nor simple, do you think that's worth it? I was leaning towards risking breakage here as my preferred option.

Reply:
I was just making sure this was a deliberate decision. I also favor the cleanest path forward (if we are allowed to break things when we believe it makes sense, which I didn't know 😄).

Reply:
It's a calculated risk and one that we generally try hard to avoid. But adding a new `abstract` predicate isn't really possible to do without potential breakage.